AttributeError in `Checkpoint::reload_objects` when `save_handler` is a callable object

## 🐛 Bug description

When using `Checkpoint` with a callable object as `save_handler`, calling `reload_objects` causes `AttributeError`. To reproduce:

```python
def test_checkpoint_simple_recovery():
    from pickle import dump
    class SaveHandler:
        def __call__(self, checkpoint, filename):
            with open(filename, 'wb') as f:
                dump(checkpoint, f)
    to_save = {'model': DummyModel()}
    h = Checkpoint(to_save, SaveHandler())
    engine = Engine(lambda e, b: None)
    engine.state = State(epoch=0, iteration=1)
    h(engine)
    to_load = {"model": DummyModel()}
    h.reload_objects(to_load, global_step=1)
```

result:
```
>       path = self.save_handler.dirname / checkpoint_fp
E       AttributeError: 'SaveHandler' object has no attribute 'dirname'
```

It attempts to access `dirname` of the handler while `dirname` is only the attribute of the `DiskSaver` not even `BaseSaveHandler`.

Also in the `Checkpoint` docstring, it states `String, method or callable class` (by callable class it means callable object?) for  `save_handler` type but in type checks we allow a callable function:

```python
if not (
    isinstance(save_handler, str)
    or isinstance(save_handler, Path)
    or callable(save_handler)
    or isinstance(save_handler, BaseSaveHandler)
):
    raise TypeError(
        "Argument `save_handler` should be a string or Path object or callable or inherit from BaseSaveHandler"
    )
```



<details>
<summary>Checklist</summary>

- [ ] `ignite/handlers/checkpoint.py`
> • In the `reload_objects` method of the `Checkpoint` class, before the line `path = self.save_handler.dirname / checkpoint_fp`, add a condition to check if `self.save_handler` is an instance of `DiskSaver`.
> • If `self.save_handler` is an instance of `DiskSaver`, keep the existing logic to get the `path`.
> • If `self.save_handler` is not an instance of `DiskSaver`, handle the checkpoint reloading for the callable object. This might involve calling the `save_handler` with the appropriate arguments to reload the checkpoint.

</details>

--- a/ignite/handlers/checkpoint.py
+++ b/ignite/handlers/checkpoint.py
@@ -670,7 +670,8 @@ class Checkpoint(Serializable):
             `DataParallel`_, method ``load_state_dict`` will applied to their internal wrapped model (``obj.module``).
 
         Note:
-            This method works only when the ``save_handler`` is of type :class:`~ignite.handlers.checkpoint.DiskSaver`.
+            This method works only when the ``save_handler`` is of types string,
+            :class:`~pathlib.Path` or :class:`~ignite.handlers.checkpoint.DiskSaver`.
 
         .. _DistributedDataParallel: https://pytorch.org/docs/stable/generated/
             torch.nn.parallel.DistributedDataParallel.html

--- a/ignite/handlers/checkpoint.py
+++ b/ignite/handlers/checkpoint.py
@@ -669,10 +669,17 @@ class Checkpoint(Serializable):
             If ``to_load`` contains objects of type torch `DistributedDataParallel`_ or
             `DataParallel`_, method ``load_state_dict`` will applied to their internal wrapped model (``obj.module``).
 
+        Note:
+            This method works only when the ``save_handler`` is of type :class:`~ignite.handlers.checkpoint.DiskSaver`.
+
         .. _DistributedDataParallel: https://pytorch.org/docs/stable/generated/
             torch.nn.parallel.DistributedDataParallel.html
         .. _DataParallel: https://pytorch.org/docs/stable/generated/torch.nn.DataParallel.html
         """
+        if not isinstance(self.save_handler, DiskSaver):
+            raise AttributeError(
+                f"Checkpoint's `save_handler` should be of type `DiskSaver`, given {type(self.save_handler)}"
+            )
 
         global_step = filename_components.get("global_step", None)
 

--- a/tests/ignite/handlers/test_checkpoint.py
+++ b/tests/ignite/handlers/test_checkpoint.py
@@ -78,6 +78,10 @@ def test_checkpoint_wrong_input():
     with pytest.raises(TypeError, match="If `include_self` is True, then `to_save` must be mutable"):
         Checkpoint(ImmutableMapping(), lambda x: x, include_self=True)
 
+    checkpoint = Checkpoint(to_save, lambda x: x)
+    with pytest.raises(AttributeError, match="Checkpoint's `save_handler` should be of type `DiskSaver`"):
+        checkpoint.reload_objects(to_save)
+
 
 def test_save_handler_as_str(dirname):
     to_save = {"model": model}

